Event.php

```diff
@@ -11,205 +11,238 @@
     {
-        $request = $event->getRequest();
-        $response = $event->getResponse();
-        if ($request->attributes->get('id')) {
-            $id = $request->attributes->get('id');
-        } else {
-            $location = explode('/', $response->headers->get('location'));
-            $url = explode('/', $this->app->url('admin_product_product_edit', array('id' => '0')));
-            $diffs = array_values(array_diff($location, $url));
-            $id = $diffs[0];
+        if ($this->supportNewHookPoint()) {
+            return;
         }
-
-        $Product = $this->app['eccube.repository.product']->find($id);
-
-        return $Product;
+        $this->legacyEvent->onRenderProductDetailBefore($event);
     }
 
-    public function addContentOnProductEdit(FilterResponseEvent $event)
+    /**
+     * for v3.0.0 - 3.0.8.
+     *
+     * @deprecated for since v3.0.0, to be removed in 3.1
+     *
+     * @param FilterResponseEvent $event
+     */
+    public function onRenderAdminProductEditBefore(FilterResponseEvent $event)
     {
-        $app = $this->app;
-        if (!$app->isGranted('ROLE_ADMIN')) {
+        if ($this->supportNewHookPoint()) {
             return;
         }
-        $request = $event->getRequest();
-        $response = $event->getResponse();
-        $id = $request->attributes->get('id');
-
-        $html = $response->getContent();
-        $crawler = new Crawler($html);
+        $this->legacyEvent->onRenderAdminProductEditBefore($event);
+    }
 
-        $form = $app['form.factory']
-            ->createBuilder('admin_product')
-            ->getForm();
+    /**
+     * v3.0.9以降のフックポイントに対応しているのか.
+     *
+     * @return bool
+     */
+    private function supportNewHookPoint()
+    {
+        return version_compare('3.0.9', Constant::VERSION, '<=');
+    }
 
-        if ($id) {
-            $Product = $app['eccube.repository.product']->find($id);
+    /**
+     * @param Product $Product
+     *
+     * @return Collection RelatedProducts
+     */
+    private function createRelatedProductData($Product)
+    {
+        $app = $this->app;
+        if ($Product) {
+            $RelatedProducts = $app['eccube.plugin.repository.related_product']->findBy(
+                array(
+                    'Product' => $Product,
+                ));
         } else {
-            $Product = new \Eccube\Entity\Product();
+            $Product = new Product();
         }
-
-        $RelatedProducts = $app['eccube.plugin.repository.related_product']->findBy(
-            array(
-                'Product' => $Product,
-            ));
-
         $loop = 5 - count($RelatedProducts);
-        for ($i = 0; $i < $loop; $i++) {
-            $RelatedProduct = new \Plugin\RelatedProduct\Entity\RelatedProduct();
+        for ($i = 0; $i < $loop; ++$i) {
+            $RelatedProduct = new RelatedProduct();
             $RelatedProduct
-                ->setProductId($id)
+                ->setProductId($Product->getId())
```

Comment:
If Product is null, You can't get product id => errors.
Implement this case to unit test.